### PR TITLE
Add SCEP support (WIP)

### PR DESCRIPTION
**Pull request by hslatman**
This PR adds [SCEP](https://tools.ietf.org/html/rfc8894) support to step-ca.

Currently no automated tests exist yet. The implementation was tested with [micromdm/scep](https://github.com/micromdm/scep) and [certnanny/sscep](https://github.com/certnanny/sscep).

It probably needs a bit of work before it can be merged; feedback welcome 😄 

Will rebase/update this PR before the actual merge.

### CODE REVIEW

**Comment by hslatman** [comment]
(High level) instructions for testing:

Prepare the CA with RSA root and intermediate keys (no password for a bit of convenience when testing):

```bash
step certificate create TestRoot rsa-root.crt rsa-root.key --profile=root-ca  --no-password --insecure --kty=RSA
step certificate create TestIntermediate rsa-intermediate.crt rsa-intermediate.key --profile=intermediate-ca --ca=rsa-root.crt --ca-key=rsa-root.key --no-password --insecure --kty=RSA
```

I haven't tested with a root that is not RSA yet. In theory it can work, but SCEP clients may not like it. The intermediate certificate and key will be used to sign new SCEP certificates. The RFC only mentions RSA; EC crypto may work, but also depends on client support.

Configure the CA (essential fields shown only):

```json
{
        ...
        "root": "/path/to/.step/certs/rsa-root.crt",
        "crt": "/path/to/.step/certs/rsa-intermediate.crt",
        "key": "/path/to/.step/secrets/rsa-intermediate.key",
        "address": ":8443",
        "insecureAddress": ":8080",
        ...
}
```

Ensure that you've moved the certificates and keys created in the previous step to the correct location corresponding to the CA configuration. As you can see, an additional address is provided to serve "insecure" endpoints, incl. those for SCEP.

Configure a SCEP provisioner:

```json
                        {
                                "type": "SCEP",
                                "name": "somename",
                                "forceCN": true,
                                "challenge": "secret1234",
                                "claims": {
                                        "maxTLSCertDuration": "48h",
                                        "defaultTLSCertDuration": "24h"
                                 }

                        }
```

Tests have been performed with [micromdm/scep](https://github.com/micromdm/scep) and [certnanny/sscep](https://github.com/certnanny/sscep).

Using [micromdm/scep](https://github.com/micromdm/scep):

- Get the code and compile the SCEP client. Easiest options is using the Makefile: `make build`. This will create a build in the `./build` dir.
- Ensure that you remove the challenge password from the provisioner configuration above. The micromdm implementation does not seem to work correctly with it at this time.
- Run the client: `./build/scepclient-os-architecture -private-key local.key -server-url http://127.0.0.1:8090/scep/somename -debug`. This should result in a new file called `client.pem`.
- For more advanced use cases, see `./build/scepclient-os-architecture -help`.

Operations (currently) supported: `getca`, `getcaps`, `enroll`. These are all executed by the client automatically.

Using [certnanny/sscep](https://github.com/certnanny/sscep):

- Get and compile sscep according to the [instructions](https://github.com/certnanny/sscep#how-to-compile). In the next steps I assume you have a binary at ./sscep.
- use the `mkrequest` script in the repository to prepare the files required for enrolling (see example at [step 2](https://github.com/certnanny/sscep#step-2---make-certificate-request-and-key)). You can provide an ip and should provide the challenge password if configured in the provisioner: `mkrequest -ip 10.10.10.10 secret1234`
- Continue with step 3. Use `http://127.0.0.1:8080/scep/somename` as the URL: `./sscep getca -u http://127.0.0.1:8080/scep/somename -c ca.crt`. This should result in a new file `ca.crt` in the cwd. Currently step only sends the CA certificate; not chain.
- Continue with step 4: `./sscep enroll -u http://127.0.0.1:8080/scep/somename -c ca.crt -k local.key -r local.csr -l local.crt -S sha1`. Ensure that you provide sha1 (or better); `sscep` will default to MD5, which is currently not supported in the `pkcs7` library. This should result in a certificate to be available in local.crt.
- Renewals (step 5) also work: `./sscep enroll -u http://127.0.0.1:8000/scep/scep1 -c cacert.crt -K local.key -O local.crt -k new.key -r new.csr -l new.crt -S sha1`. You can use the same key or use a new one.

Operations (currently) supported: `getca`, `getcaps`, `enroll`.

File: scep/authority.go [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return degenerate, nil
+}
+
+// Interface guards
+var (
+	_ Interface = (*Authority)(nil)
+)
```
Comment (maraino) [inline scep/authority.go:511]: See my previous comment in the provisioner.

File: scep/certificate.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+	_, swapped, err := db.CmpAndSwap(certTable, []byte(id), nil, certB)
+	switch {
+	case err != nil:
+		return nil, fmt.Errorf("%w: error storing certificate", err)
+	case !swapped:
+		return nil, fmt.Errorf("error storing certificate; " +
+			"value has changed since last read")
+	default:
+		return cert, nil
```
Comment (maraino) [inline scep/certificate.go:64]: This will be a method in the `scep.DB` interface

File: scep/certificate.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+// 		return nil, fmt.Errorf("certificate %s not found", id)
+// 	} else if err != nil {
+// 		return nil, fmt.Errorf("error loading certificate")
+// 	}
+// 	var cert certificate
+// 	if err := json.Unmarshal(b, &cert); err != nil {
+// 		return nil, fmt.Errorf("%w: error unmarshaling certificate", err)
+// 	}
+// 	return &cert, nil
+// }
```
Comment (maraino) [inline scep/certificate.go:80]: Remove me before finishing the PR

**Review by maraino — CHANGES_REQUESTED** [review]
Hi @hslatman I've done the first pass on this. It looks pretty good. And thanks again.

File: authority/authority.go [hunk 4]
```diff
@@ -338,6 +342,55 @@ func (a *Authority) init() error {
...
+				SigningKey: a.config.IntermediateKey,
+				Password:   []byte(a.config.Password),
+			})
+			if err != nil {
+				return err
+			}
+
+			if km, ok := a.keyManager.(kmsapi.Decrypter); ok {
+				options.Decrypter, err = km.CreateDecrypter(&kmsapi.CreateDecrypterRequest{
+					DecryptionKey: a.config.IntermediateKey,
```
[the remark this answers is not in the document]
Comment (hslatman) [reply authority/authority.go]: Right now SCEP indeed requires the intermediate key for decryption. A SCEP client uses a CA certificate, in this case the intermediate certificate, to encrypt messages. The SCEP authority needs the key to decrypt those messages.

The CA certificate does not need to be the same as the intermediate; it can be a different certificate/key combination. SCEP is normally performed over HTTP, meaning there's no reason to have an HTTPS connection with corresponding CA. The current configuration is kind of a convenient way to run SCEP. I think it may even be good idea to totally separate the "TLS CA" from the "SCEP CA", but I'm not an expert in this area.

I'm not sure if the key should be in the provisioner directly. I do think it's a good thing to make it possible to control the key that's used by the SCEP authority via the provisioner. The actual key should IMO exist in a KMS, but I think we agree on that part already.

File: authority/provisioner/scep.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+// GetCapabilities returns the CA capabilities
+func (s *SCEP) GetCapabilities() []string {
+	return s.Capabilities
+}
+
+// Interface guards
+var (
+	_ Interface = (*SCEP)(nil)
+	//_ scep.Provisioner = (*SCEP)(nil)
+)
```
[the remark this answers is not in the document]
Comment (hslatman) [reply authority/provisioner/scep.go]: It is, indeed. They are referred to as "interface guards". They are useful to validate that certain interfaces are implemented at runtime in cases where there's some flexibility through configuration. I learned about them from Caddy, where they're used to ensure that (3rd party) modules implement interfaces and can be checked at compile time instead of at runtime.

You're right that they're not really useful here. I used them during development to highlight the functions I had to implement. I've removed them now.

File: ca/ca.go [hunk 6]
```diff
@@ -293,3 +356,14 @@ func (ca *CA) getTLSConfig(auth *authority.Authority) (*tls.Config, error) {
...
+// func dumpRoutes(mux chi.Routes) {
+// 	// helpful routine for logging all routes //
+// 	walkFunc := func(method string, route string, handler http.Handler, middlewares ...func(http.Handler) http.Handler) error {
+// 		fmt.Printf("%s %s\n", method, route)
+// 		return nil
+// 	}
+// 	if err := chi.Walk(mux, walkFunc); err != nil {
+// 		fmt.Printf("Logging err: %s\n", err.Error())
+// 	}
+// }
```
[the remark this answers is not in the document]
Comment (hslatman) [reply ca/ca.go]: I could remove this, but it was existing commented code that I moved to a function for convenience.

File: scep/authority.go [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+	return a.getLinkExplicit(provName, abs, baseURL, inputs...)
+}
+
+// getLinkExplicit returns an absolute or partial path to the given resource and a base
+// URL dynamically obtained from the request for which the link is being calculated.
+func (a *Authority) getLinkExplicit(provisionerName string, abs bool, baseURL *url.URL, inputs ...string) string {
+
+	// TODO: do we need to provide a way to provide a different suffix/base?
+	// Like "/cgi-bin/pkiclient.exe"? Or would it be enough to have that as the name?
+	link := fmt.Sprintf("/%s", provisionerName)
```
[the remark this answers is not in the document]
Comment (hslatman) [reply scep/authority.go:118]: It currently is formatted as what you mention here, but the RFC mentions "/cgi-bin/pkiclient.exe" explicitly and I've seen examples of other paths. My comment is about making it possible to configure an optional suffix, so that the full path results in /scep/somename/cgi-bin/pkiclient.exe (or a different suffix).

File: scep/authority.go [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+		if err := a.DecryptPKIEnvelope(ctx, msg); err != nil {
+			return nil, err
+		}
+		csr = msg.CSRReqMessage.CSR
+	}
+
+	// Template data
+	data := x509util.NewTemplateData()
+	data.SetCommonName(csr.Subject.CommonName)
+	data.SetSANs(csr.DNSNames)
```
[the remark this answers is not in the document]
Comment (hslatman) [reply scep/authority.go]: I've removed the additional calls to data.SetX, because they seem redundant. The CSR is passed to the authority via SignCSR.

File: ca/ca.go [hunk 9]
```diff
@@ -143,16 +149,34 @@ func (ca *CA) Init(config *authority.Config) (*CA, error) {
...
-		}
-		if err := chi.Walk(mux, walkFunc); err != nil {
-			fmt.Printf("Logging err: %s\n", err.Error())
-		}
-	*/
+	scepPrefix := "scep"
+	scepAuthority, err := scep.New(auth, scep.AuthorityOptions{
+		Service:  auth.GetSCEPService(),
+		Backdate: *config.AuthorityConfig.Backdate,
+		DB:       auth.GetDatabase().(nosql.DB),
```
[the remark this answers is not in the document]
Comment (hslatman) [reply ca/ca.go]: I've created a scep.DB. The only thing that's stored currently is a new certificate; no new functionalities required, unless more SCEP functions are to be implemented.

File: scep/certificate.go [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+)
+
+// CertOptions options with which to create and store a cert object.
+type CertOptions struct {
+	Leaf          *x509.Certificate
+	Intermediates []*x509.Certificate
+}
+
+func newCert(db DB, ops CertOptions) error {
+	err := db.StoreCertificate(ops.Leaf)
```
Comment (maraino) [inline scep/certificate.go:16]: The `authority.Sign` method will store the certificate to, why do we want to store it again?

File: ca/ca.go [hunk 11]
```diff
@@ -143,16 +149,33 @@ func (ca *CA) Init(config *authority.Config) (*CA, error) {
...
+		DNS:     dns,
+		Prefix:  scepPrefix,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating SCEP authority")
+	}
+	scepRouterHandler := scepAPI.New(scepAuthority)
+	mux.Route("/"+scepPrefix, func(r chi.Router) {
+		scepRouterHandler.Route(r)
+	})
```
[the remark this answers is not in the document]
Comment (maraino) [reply ca/ca.go:165]: I've just seen that scep.DB only has `StoreCertificate(crt *x509.Certificate) error`, there's no need to "cast" nor check that `db.AuthDB` from `GetDatabase()` implements it, because StoreCertificate is a method in db.AuthDB. 

But see my comment below, I don't think you need to use StoreCertificate.

**Review by maraino — CHANGES_REQUESTED** [review]
I've added a few more comments.

File: scep/authority.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (maraino) [reply scep/authority.go:118]: I meant that there's no need to go through all the sprintf parsing for this, totally unnecessary performance improvement, though.

File: scep/authority.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (maraino) [reply scep/authority.go]: But the default template will ignore the other SANs if you don't add them here. We apply a template to the CSR to only include in the final certificates things that we want.

[the remark this answers is not in the document]
Comment (hslatman) [reply scep/authority.go]: I've included more data now, based on this: https://github.com/smallstep/cli/blob/a5f85865a85f5ebb33347fd7350d1ebc7d9c870a/command/certificate/sign.go#L379

File: scep/certificate.go (hunk 10 above)
Comment (hslatman) [reply scep/certificate.go]: Ah, I wasn't aware of that and (incorrectly) assumed I needed the same logic as for ACME. I've removed the database from the implementation.

In SCEP there exists a flow in which a certificate is requested but not immediately created and signed. In this case, the client should start polling for its certificate. Approving the certificate is performed as a separate process (i.e. manual authorization). This flow likely requires some state to be kept.

The current state of the integration of SCEP doesn't support this and I also think that for step-ca this workflow might not be that interesting to support. If it has to be supported, a method for interacting with step-ca should be created that can used to approve a SCEP request. This could be part of step-cli.

File: scep/authority.go [hunk 12]
```diff
@@ -0,0 +1,473 @@
...
+	return a.getLinkExplicit(provName, abs, baseURL, inputs...)
+}
+
+// getLinkExplicit returns an absolute or partial path to the given resource and a base
+// URL dynamically obtained from the request for which the link is being calculated.
+func (a *Authority) getLinkExplicit(provisionerName string, abs bool, baseURL *url.URL, inputs ...string) string {
+
+	// TODO: do we need to provide a way to provide a different suffix?
+	// Like "/cgi-bin/pkiclient.exe"? Or would it be enough to have that as the name?
+	link := "/" + provisionerName
```
[the remark this answers is not in the document]
Comment (maraino) [reply scep/authority.go:113]: But if the non-absolute link is like `/{provisionerID}`, that I think is the case in the `Handler.Route` method, then we should not escape those brackets, as the router is not going to be able to handle them. I would use a different method for route links and other links.

**Review by maraino — CHANGES_REQUESTED** [review]
Hi @hslatman I had some time to take a second look to this. I've added a few extra comments, my main concern is the `challenge` secret, that I think right now it's public, and I'm not sure if it should be.

File: scep/authority.go [hunk 13]
```diff
@@ -0,0 +1,473 @@
...
+		if u.Scheme == "" {
+			u.Scheme = "http"
+		}
+
+		// If no Host is set, then use the default (first DNS attr in the ca.json).
+		if u.Host == "" {
+			u.Host = a.dns
+		}
+
+		u.Path = a.prefix + link
```
[the remark this answers is not in the document]
Comment (hslatman) [reply scep/authority.go:132]: I think it works as you would expect, indeed. The /scep prefix is still in use when _not_ using the absolute URL.

File: scep/authority.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (hslatman) [reply scep/authority.go:113]: Currently only the name of the provisioner is used as part of the path. 

What exactly are you looking for to change? Currently URLs are mounted at the `/scep` prefix with the name of the provisioner behind it (`/provisionerName`) and then the standard SCEP paths.

File: authority/provisioner/scep.go [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+// SCEP is the SCEP provisioner type, an entity that can authorize the
+// SCEP provisioning flow
+type SCEP struct {
+	*base
+	Type string `json:"type"`
+	Name string `json:"name"`
+
+	ForceCN           bool     `json:"forceCN,omitempty"`
+	ChallengePassword string   `json:"challenge,omitempty"`
```
[the remark this answers is not in the document]
Comment (hslatman) [reply authority/provisioner/scep.go:18]: I hadn't used that function of the Step CLI yet, but you're most probably right. When I tried this however, I didn't get the output that I would expect for the SCEP provisioner; only the default JWK one was shown so far, but I guess that's due to my local Step CLI version not understanding SCEP yet? Although the code in that repo seems to show it should show the JSON, so I'm not fully sure why I didn't get the SCEP as JSON output.

The challenge should indeed be considered secret. As it is used now, it's a shared secret amongst all SCEP clients connecting to this CA. This is in line with the RFC. Of course it would be nice to have multiple secrets, perhaps even unique per client, but that would require additional storage or a different way to provide these to Step.

Anyway, I made a relatively small change that redacts the output in the provisioner `Init` function. This seemed to be the solution that required the least amount of changes and seems effective so far.

**Review by maraino — APPROVED** [review]
Hi @hslatman, 

The PR looks good, but I'm not gonna merge it yet.

I need to figure out a better way to handle the challenge secret. As it is right now, we will lose the current challenge if we add a new provisioner using `step ca provisioner add. Perhaps I will add support to load the challenge from a file too.

I also would like to write to the logs some error conditions. I've tested, and it was hard to figure out why the requests were failing.

@hslatman: PKIOperation requests, even with the errors like "invalid challenge" or the use of an RSA key of 1024 bits were returning 200, is that expected?.`

I'll be adding commits to this PR to solve these things. Feel free to add new things too.

**Comment by hslatman** [comment]
> Hi @hslatman,
> 
> The PR looks good, but I'm not gonna merge it yet.
> 
> I need to figure out a better way to handle the challenge secret. As it is right now, we will lose the current challenge if we add a new provisioner using `step ca provisioner add. Perhaps I will add support to load the challenge from a file too.

Ah, this sounds like it has something to do with the reload functionality? I had some doubts about that part, lately. A file may indeed be a relative simple way for providing (multiple) challenge secrets. 

> I also would like to write to the logs some error conditions. I've tested, and it was hard to figure out why the requests were failing.

Sounds good.
 
> @hslatman: PKIOperation requests, even with the errors like "invalid challenge" or the use of an RSA key of 1024 bits were returning 200, is that expected?.`

Yes, I would say that works as expected. When trying it with the (current) client implementations and returning something different than 200, this resulted in the clients not handling the responses correctly. The actual errors are (when it's possible to do so) packaged inside of a PKCS7 structure (see [pkiStatus](https://tools.ietf.org/html/rfc8894#section-3.2.1.3)). There's some room for improvement, though. I can look into that when adding the tests. Besides that, there may be clients that behave differently, of course. 

Regarding the 1024 bits: technically I don't think it's an issue for SCEP to use weak keys (as long as they're RSA keys; or encrypted using alternative means, as described [here](https://tools.ietf.org/html/rfc8894#section-3.1)); I think I've even seen examples using 1024 bits. I guess we could add an additional check for this when the Decrypter configuration is validated. I didn't test with 1024 bits yet, so I gather that this resulted in an error when you tested it?

> I'll be adding commits to this PR to solve these things. Feel free to add new things too.

Sounds good. I'll keep an eye on your changes to not do double work. I'll be adding test cases soon

**Comment by maraino** [comment]
> Regarding the 1024 bits: technically I don't think it's an issue for SCEP to use weak keys (as long as they're RSA keys; or encrypted using alternative means, as described here); I think I've even seen examples using 1024 bits. I guess we could add an additional check for this when the Decrypter configuration is validated. I didn't test with 1024 bits yet, so I gather that this resulted in an error when you tested it?

sscep's mkrequest creates by default 1024bits keys, that we do not allow. I don't think this is important right now, as in the future if we have users with this problem, then we can make this configurable.

> Ah, this sounds like it has something to do with the reload functionality? I had some doubts about that part, lately. A file may indeed be a relative simple way for providing (multiple) challenge secrets.

I've just tested the re-load and works fine, we always re-read the file that has the password, so we should be good. And `step ca provisioner add` should not be a problem, because it also uses the file. 

To summarize, I think the only thing that we should do is to add errors to the logs to actually have a way to know what is going on.

**Comment by hslatman** [comment]
> sscep's mkrequest creates by default 1024bits keys, that we do not allow. I don't think this is important right now, as in the future if we have users with this problem, then we can make this configurable.

Good catch. I had changed this locally at some point to get around that particular error 😅. I have added an additional option and public key length validator to allow for keys shorter than 2048 now. 

> To summarize, I think the only thing that we should do is to add errors to the logs to actually have a way to know what is going on.

I've improved this and some more errors should now be visible in the logs. There may be some more cases to log explicitly; will check to find some more when implementing tests.

**Review by maraino — APPROVED** [review]
I've been testing this again, and the errors are nicely logged. Thanks.
I'll talk tomorrow with the team about merging this now.

**Comment by brotherdust** [comment]
CA won't start with SCEP enabled. I suspect it's because my certs are EC instead of RSA.
Here's the error:
```
ca_1  | 2021/06/11 15:17:49 reloading ...
ca_1  | 2021/06/11 15:17:49 Reload failed because the CA with new configuration could not be initialized.
ca_1  | 2021/06/11 15:17:49 Continuing to run with the original configuration.
ca_1  | 2021/06/11 15:17:49 You can force a restart by sending a SIGTERM signal and then restarting the step-ca.
ca_1  | 2021/06/11 15:17:49 error reloading server: decryptionKey is not a crypto.Decrypter
ca_1  | github.com/smallstep/certificates/kms/softkms.(*SoftKMS).CreateDecrypter
ca_1  |         /step-ca/kms/softkms/softkms.go:177
ca_1  | github.com/smallstep/certificates/authority.(*Authority).init
ca_1  |         /step-ca/authority/authority.go:372
ca_1  | github.com/smallstep/certificates/authority.New
ca_1  |         /step-ca/authority/authority.go:88
ca_1  | github.com/smallstep/certificates/ca.(*CA).Init
ca_1  |         /step-ca/ca/ca.go:116
ca_1  | github.com/smallstep/certificates/ca.New
ca_1  |         /step-ca/ca/ca.go:94
ca_1  | github.com/smallstep/certificates/ca.(*CA).Reload
ca_1  |         /step-ca/ca/ca.go:313
ca_1  | github.com/smallstep/certificates/ca.StopReloaderHandler
ca_1  |         /step-ca/ca/signal.go:59
ca_1  | runtime.goexit
ca_1  |         /usr/local/go/src/runtime/asm_amd64.s:1371
ca_1  | error reloading ca
ca_1  | github.com/smallstep/certificates/ca.(*CA).Reload
ca_1  |         /step-ca/ca/ca.go:321
ca_1  | github.com/smallstep/certificates/ca.StopReloaderHandler
ca_1  |         /step-ca/ca/signal.go:59
ca_1  | runtime.goexit
ca_1  |         /usr/local/go/src/runtime/asm_amd64.s:1371
```

**Comment by maraino** [comment]
@brotherdust Yes, currently you need at least the intermediate to be an RSA certificate because (currently) we are using it to decrypt the SCEP message, and you cannot do that with an EC key.

**Comment by brotherdust** [comment]
@maraino ,
Thank you for responding so quickly. It's welcome news that I don't have to generate a whole new root. Would you mind answering a couple more questions?
- Since the intermediate is currently EC, should I just make another intermediate signed with RSA and just use it for SCEP only?
- What about making an RSA-signed subordinate to the current EC-signed intermediate?
- For that matter, what about spinning up a second instance (with the previously-mentioned RSA subordinate) with the RA function enabled, pointing to the EC-signed intermediate?

I realize this is a complex topic and I might be misunderstanding a few things; any insight you have would be appreciated.

On a side note, is the documentation for this function being actively worked on by anyone? I'd like to contribute to the docs as a way of saying thank you.

**Comment by maraino** [comment]
> Since the intermediate is currently EC, should I just make another intermediate signed with RSA and just use it for SCEP only?

You can, but right now that specific CA will always sign certificates with the RSA certificate, at some point we might add a custom key for scep, but this is not yet supported.

> What about making an RSA-signed subordinate to the current EC-signed intermediate?

You can, but you need an EC intermediate with max length of at least 1, and the one created by `step ca init` has max length set to 0.  And of course, you will need a root with max length set to 2. You can use `step certificates create` with custom templates to generate those.

These are the default templates, changing maxPathLen will allow this
https://github.com/smallstep/crypto/blob/a491387f59350664ea5bf7972b315a596a81513d/x509util/templates.go#L149-L170

> For that matter, what about spinning up a second instance (with the previously-mentioned RSA subordinate) with the RA function enabled, pointing to the EC-signed intermediate?

I think this makes perfect sense, but the SCEP client needs to support certificates signed by an EC certificate. I haven't tested this configuration. 

```
SCEP CLIENT <---> SCEP RA <--|
                             |
OTHER CLIENT ---- (RA) ---  CA
```

Or intermediates sharing the same root:
```
SCEP CLIENT <---> SCEP CA (with RSA intermediate)
                             
OTHER CLIENT <--> CA (with EC intermedate)
```

Both configurations are possible, but as I said, the hardware support for EC signed certificates/intermediates will be decisive for this.

**Comment by maraino** [comment]
> On a side note, is the documentation for this function being actively worked on by anyone? I'd like to contribute to the docs as a way of saying thank you.

That would be awesome, cc @devadvocado

**Comment by brotherdust** [comment]
@maraino
Success! I got SCEP to work with Mikrotik RouterOS 6.48 as a client.
EC-signed root --> RSA-signed "compatibility" intermediate --> SCEP

It would have been much better to allow for a custom cert for the SCEP provisioner, as now I have two separate instances; but it works, and I think that's great!

Provided @devadvocado is welcome docs contributions, I can get that done as time permits.

**Comment by ajaymdesai** [comment]
@brotherdust  Can you tell me how you step SCEP. I'm trying it with my MikroTik hEXs. How do I create the RSA intermediate and add the SCEP provisioner.
